cmd/compare/compare.go

```diff
@@ -1,37 +1,262 @@
+		return nil
+	}
+
+	cmd.Flags().StringSliceVarP(&options.ResourceDirs, "resource", "", []string{}, "User configuration directory to read from")
+
+	err = cmd.MarkFlagRequired("resource")
+	if err != nil {
+		return nil
+	}
+
+	cmd.Flags().BoolVarP(&options.ExactMatchOnly, "exact-match-only", "", false, "Return early by determining if both sets are exact match")
+
+	return cmd
+}
+
+func (o compareOptions) validate() error {
+	for _, dir := range o.ReferenceDirs {
+		if !util.IsDirectory(dir) {
+			return errors.New("all Reference paths must be a directory")
+		}
+	}
+
+	for _, dir := range o.ResourceDirs {
+		if !util.IsDirectory(dir) {
+			return errors.New("all Resource paths must be a directory")
+		}
+	}
+
+	return nil
+}
+
+func (o compareOptions) run() {
+	slog.Info("preparing resources")
+
+	var uListResources []unstructured.Unstructured
+
+	uListResources = readK8sResourcesFromDir(o.ResourceDirs, uListResources)
+	uListResources = getResourceFromPolicyIfAny(uListResources)
+
+	slog.Info("preparing reference")
+
+	var uListReference []unstructured.Unstructured
+
+	uListReference = readK8sResourcesFromDir(o.ReferenceDirs, uListReference)
+	uListReference = getResourceFromPolicyIfAny(uListReference)
+
+	// short circuit. Useful for ACM vs ZTP cases
+	eMatch := equalUnstructuredList(uListResources, uListReference)
+
+	if o.ExactMatchOnly {
+		slog.Info("exiting early")
+
+		if eMatch {
+			os.Exit(0)
+		}
+
+		os.Exit(1)
+	}
+
+	_ = k8sdiff.DiffProgram{
+		Exec:      nil,
+		IOStreams: genericclioptions.IOStreams{},
+	}
+}
+
+func getResourceFromPolicyIfAny(uList []unstructured.Unstructured) []unstructured.Unstructured {
+	// Extract the main CR if policy
+	var uListWithoutP []unstructured.Unstructured
+
+	for _, curUnstructured := range uList {
+		if curUnstructured.GetKind() == "Policy" {
+			policy := policyv1.Policy{}
+
+			err := runtime.DefaultUnstructuredConverter.FromUnstructured(curUnstructured.Object, &policy)
+			if err != nil {
+				slog.Warn("invalid Policy CR")
+
+				continue
+			}
+
+			uListWithoutP = append(uListWithoutP, getObjectTemplates(policy)...)
+
+			continue
 		}
 
-		return fmt.Errorf("both paths must be a directory")
-	},
-	Run: func(cmd *cobra.Command, args []string) {
-		fmt.Println("compare called")
-	},
+		uListWithoutP = append(uListWithoutP, curUnstructured)
+	}
+
+	return uListWithoutP
+}
+
+func readK8sResourcesFromDir(curDir []string, uList []unstructured.Unstructured) []unstructured.Unstructured {
+	for _, d := range curDir {
+		files, _ := util.GetFileNames(d)
+		for _, f := range files {
+			u := yamlToUnstructured(f)
+			uList = append(uList, *u)
+		}
+	}
+
+	return uList
 }
 
-func init() {
-	rootCmd.AddCommand(compareCmd)
+func yamlToUnstructured(file string) *unstructured.Unstructured {
+	yFile, _ := os.ReadFile(file)
+	newUnstructured := &unstructured.Unstructured{Object: map[string]interface{}{}}
+
+	if err := yaml.Unmarshal(yFile, &newUnstructured.Object); err != nil {
+		os.Exit(1)
```

Comment:
We should start working in failure mode error propagation from the start. At a minimum we need a reasonable error message here and should consider propagating the error up to the calling context.

Reply:
Updated to handle the error better + warning msg